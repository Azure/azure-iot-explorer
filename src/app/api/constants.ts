/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
export enum HTTP_OPERATION_TYPES {
    Delete = 'DELETE',
    Get = 'GET',
    Patch = 'PATCH',
    Post = 'POST',
    Put = 'PUT'
}

export const APPLICATION_JSON = 'application/json';
export const ERROR_TYPES = {
    HTTP: 'http'
};
