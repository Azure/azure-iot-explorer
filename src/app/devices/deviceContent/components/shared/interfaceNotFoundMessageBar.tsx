/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { MessageBar, MessageBarType } from 'office-ui-fabric-react/lib/MessageBar';
import { MessageBarButton } from 'office-ui-fabric-react/lib/Button';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import '../../../../css/_interfaceNotFoundMessageBar.scss';

export interface InterfaceNotFoundMessageBoxProps {
    settingsVisibleToggle: (visible: boolean) => void;
}

export default class InterfaceNotFoundMessageBox extends React.Component<InterfaceNotFoundMessageBoxProps> {

    public render() {
        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                    <>
                    <MessageBar
                        messageBarType={MessageBarType.error}
                        actions={
                            <MessageBarButton
                                className="configure-button"
                                onClick={this.handleConfigure}
                            >
                                    {context.t(ResourceKeys.deviceInterfaces.command.configure)}
                            </MessageBarButton>}
                    >
                        {context.t(ResourceKeys.notifications.getInterfaceModelOnError)}
                    </MessageBar>
                    </>
                )}
            </LocalizationContextConsumer>
        );
    }

    private readonly handleConfigure = () => {
        this.props.settingsVisibleToggle(true);
}
}
