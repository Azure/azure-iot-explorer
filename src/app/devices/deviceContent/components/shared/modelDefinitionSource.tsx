/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Redirect } from 'react-router-dom';
import { Stack } from 'office-ui-fabric-react/lib/Stack';
import { ActionButton } from 'office-ui-fabric-react/lib/Button';
import { Label } from 'office-ui-fabric-react/lib/Label';
import { ROUTE_PARTS, ROUTE_PARAMS } from '../../../../constants/routes';
import { useLocalizationContext } from '../../../../shared/contexts/localizationContext';
import { REPOSITORY_LOCATION_TYPE } from '../../../../constants/repositoryLocationTypes';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import '../../../../css/_modelDefinitionSource.scss';

interface ModelDefinitionSourceViewProps {
    source: REPOSITORY_LOCATION_TYPE;
}

const getModelDefinitionSourceResourceKeys = (source: REPOSITORY_LOCATION_TYPE) => {
    if (Object.values(REPOSITORY_LOCATION_TYPE).includes(source)) {
        // tslint:disable-next-line:no-any
        return (ResourceKeys.modelRepository.types as any)[source.toLowerCase()].label;
    }
    return ResourceKeys.modelRepository.types.notAvailable;
};

export const ModelDefinitionSourceView: React.FC<ModelDefinitionSourceViewProps> = props => {
    const { t } = useLocalizationContext();
    const [redirectToModelRepositories, setRedirectToModelRepositories] = React.useState<boolean>(false);

    const onConfigureClick = () => {
        setRedirectToModelRepositories(true);
    };

    if (redirectToModelRepositories) {
        return <Redirect to={`/${ROUTE_PARTS.HOME}/${ROUTE_PARTS.MODEL_REPOS}?${ROUTE_PARAMS.NAV_FROM}`} />;
    }

    return (
        <Stack horizontal={true}>
            <Stack.Item align="start">
                <Label>{t(ResourceKeys.deviceInterfaces.columns.source)}: {t(getModelDefinitionSourceResourceKeys(props.source))}</Label>
            </Stack.Item>

            <Stack.Item align="center">
                <ActionButton
                    className="configure-button"
                    onClick={onConfigureClick}
                >
                        {t(ResourceKeys.deviceInterfaces.command.configure)}
                </ActionButton>
            </Stack.Item>
        </Stack>
    );
};
