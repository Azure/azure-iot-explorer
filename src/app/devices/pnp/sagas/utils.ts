import { ModelIdCasingNotMatchingException } from '../../../shared/utils/exceptions/modelIdCasingNotMatchingException';
import { ModelDefinition } from '../../../api/models/modelDefinition';
import { ModelRepositoryConfiguration } from '../../../shared/modelRepository/state';
import { REPOSITORY_LOCATION_TYPE } from '../../../constants/repositoryLocationTypes';

export const getSplitInterfaceId = (fullName: string) => {
    // when component definition is inline, interfaceId is compose of parent file name and inline schema id concatenated with a slash
    return fullName.split('/');
};

export const getFlattenedModel = (model: ModelDefinition, splitInterfaceId: string[]) => {
    if (splitInterfaceId.length === 1) {
        return model;
    }
    else {
        // for inline component, the flattened model is defined under contents array's with matching schema @id
        const components = model.contents.filter((content: any) => // tslint:disable-line: no-any
            content['@type'] === 'Component' && typeof content.schema !== 'string' && content.schema['@id'] === splitInterfaceId[1]);
        return components[0];
    }
};

export const checkModelIdCasing = (model: ModelDefinition, id: string) => {
    if (model['@id'] !== id) {
        throw new ModelIdCasingNotMatchingException();
    }
};

export const getLocationSettingValue = (locations: ModelRepositoryConfiguration[], type: REPOSITORY_LOCATION_TYPE): string => {
    const filteredValue = locations.filter(location => location.repositoryLocationType === type)?.[0]?.value || '';
    return filteredValue.replace(/\/$/, ''); // remove trailing slash
};

export const getDmrParams = (path: string, interfaceId: string): {folderPath: string, fileName: string} => {
    // convert dtmi name to follow drm convention
    // for example: dtmi:com:example:Thermostat;1 -> dtmi/com/example/thermostat-1.json
    const fullPath = `${path.replace(/\/$/, '')}/${interfaceId.toLowerCase().replace(/:/g, '/').replace(';', '-')}.json`;
    // folderPath will be converted to for example: ${path}/dtmi/com/example, and fileName will be thermostat-1.json
    return {folderPath: fullPath.substring(0, fullPath.lastIndexOf('/')), fileName: fullPath.substring(fullPath.lastIndexOf('/') + 1, fullPath.length)};
};
