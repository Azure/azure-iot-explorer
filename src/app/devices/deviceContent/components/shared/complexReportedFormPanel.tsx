/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Label } from 'office-ui-fabric-react/lib/Label';
import { IconButton } from 'office-ui-fabric-react/lib/Button';
import { Spinner, SpinnerSize } from 'office-ui-fabric-react/lib/Spinner';
import Form from 'react-jsonschema-form';
import { fabricWidgets, fabricFields } from '../../../../jsonSchemaFormFabricPlugin';
import { ObjectTemplate } from '../../../../jsonSchemaFormFabricPlugin/fields/objectTemplate';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { Exception } from '../../../../shared/utils/exceptions/exception';
import { ParsedJsonSchema } from '../../../../api/models/interfaceJsonParserOutput';
import { twinToFormDataConverter } from '../../../../shared/utils/twinAndJsonSchemaDataConverter';
import { CLOSE } from '../../../../constants/iconNames';
import { PropertyContent } from '../../../../api/models/modelDefinition';
import ErrorBoundary from '../../../errorBoundary';
import { getLocalizedData } from '../../../../api/dataTransforms/modelDefinitionTransform';

const EditorPromise = import('react-monaco-editor');
const Editor = React.lazy(() => EditorPromise);

export interface ReportedFormDataProps {
    showPanel: boolean;
    formData: any; // tslint:disable-line:no-any
    schema: ParsedJsonSchema;
    modelDefinition: PropertyContent;
}

export interface ReportedFormActionProps {
    handleDismiss: () => void;
}

export interface ReportedFormState {
    formData: any; // tslint:disable-line:no-any
    parseMapTypeError?: Exception;
}

export default class ComplexReportedFormPanel extends React.Component<ReportedFormDataProps & ReportedFormActionProps, ReportedFormState> {
    constructor(props: ReportedFormDataProps & ReportedFormActionProps) {
        super(props);

        const { formData, schema } = this.props;
        const twinData = twinToFormDataConverter(formData, schema);
        this.state = {
            formData: twinData.formData,
            parseMapTypeError: twinData.error
        };
    }

    public render() {
        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                    <dialog open={this.props.showPanel} role="dialog">
                        <IconButton
                            className="close-dialog-icon"
                            iconProps={{ iconName: CLOSE }}
                            onClick={this.props.handleDismiss}
                        />
                        {this.createTitle(context)}
                        {this.createForm(context)}
                    </dialog>
                )}
            </LocalizationContextConsumer>
        );
    }

    private readonly createTitle = (context: LocalizationContextInterface) => {
        const displayName = getLocalizedData(this.props.modelDefinition.displayName);
        return (
            <div className="panel-title">
                <h1>{context.t(ResourceKeys.deviceSettings.panel.title)}</h1>
                <Label>{this.props.modelDefinition.name} ({displayName ? displayName : '--'})</Label>
            </div>
        );
    }
    private readonly createForm = (context: LocalizationContextInterface) => {
        if (this.state.parseMapTypeError || !this.props.schema) { // Not able to parse interface definition, render raw json editor instead
            return this.createJsonEditor(context);
        }
        else {
            return (
                <ErrorBoundary error={context.t(ResourceKeys.errorBoundary.text)}>
                    <Form
                        formData={this.state.formData}
                        liveValidate={true}
                        schema={this.props.schema as any} // tslint:disable-line: no-any
                        showErrorList={false}
                        uiSchema={{'ui:description': this.props.schema.description, 'ui:disabled': true}}
                        widgets={fabricWidgets}
                        {...fabricFields}
                        ObjectFieldTemplate={ObjectTemplate}
                    >
                        <br/>
                    </Form>
                </ErrorBoundary>
            );
        }
    }

    private readonly createJsonEditor = (context: LocalizationContextInterface) => {
        return (
            <form>
                <Label>{context.t(ResourceKeys.deviceProperties.editor.label, {schema: this.getSettingSchema()})}</Label>
                <div className="monaco-editor">
                    <React.Suspense fallback={<Spinner title={'loading'} size={SpinnerSize.large} />}>
                        <Editor
                            language="json"
                            options={{
                                automaticLayout: true,
                                readOnly: true
                            }}
                            height="70vh"
                            value={JSON.stringify(this.state.formData, null, '\t')}
                        />
                    </React.Suspense>
                </div>
            </form>
        );
    }

    private readonly getSettingSchema = () => {
        const { modelDefinition } = this.props;
        return typeof modelDefinition.schema === 'string' ?
            modelDefinition.schema :
            modelDefinition.schema['@type'];
    }
}
