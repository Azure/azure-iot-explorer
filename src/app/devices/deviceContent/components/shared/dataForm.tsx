/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { PrimaryButton, Dialog, DialogFooter, DefaultButton } from 'office-ui-fabric-react';
import Form from 'react-jsonschema-form';
import Editor from 'react-monaco-editor';
import { fabricWidgets, fabricFields } from '../../../../jsonSchemaFormFabricPlugin';
import { ObjectTemplate } from '../../../../jsonSchemaFormFabricPlugin/fields/objectTemplate';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { SUBMIT, CODE } from '../../../../constants/iconNames';
import { Exception } from '../../../../shared/utils/exceptions/exception';
import { ParsedJsonSchema } from '../../../../api/models/interfaceJsonParserOutput';
import { dataToTwinConverter, twinToFormDataConverter } from '../../../../shared/utils/twinAndJsonSchemaDataConverter';
import ErrorBoundary from '../../../../devices/errorBoundary';
import LabelWithTooltip from '../../../../shared/components/labelWithTooltip';
import '../../../../css/_dataForm.scss';

const payloadRef = React.createRef<Editor>();

export interface DataFormDataProps {
    formData: any;  // tslint:disable-line:no-any
    interfaceName: string;
    settingSchema: ParsedJsonSchema;
    buttonText: string;
    schema: string; // the schema as defined in model definition
}

export interface DataFormActionProps {
    handleSave: (twin: any) => () => void; // tslint:disable-line:no-any
    craftPayload: (payload: object) => object;
}

export interface DataFormState {
    formData: any; // tslint:disable-line:no-any
    originalFormData: any; // tslint:disable-line:no-any
    stringifiedFormData: string;
    parseMapTypeError?: Exception;
    showPayloadDialog?: boolean;
    payloadPreviewData?: any; // tslint:disable-line: no-any
}
export default class DataForm extends React.Component<DataFormDataProps & DataFormActionProps, DataFormState> {
    constructor(props: DataFormDataProps & DataFormActionProps) {
        super(props);

        const { formData, settingSchema } = this.props;
        const twinData = twinToFormDataConverter(formData, settingSchema);
        this.state = {
            formData: twinData.formData,
            originalFormData: twinData.formData,
            parseMapTypeError: twinData.error,
            stringifiedFormData: JSON.stringify(twinData.formData, null, '\t')
        };
    }

    public render() {
        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                    <>
                        {this.createForm(context)}
                        {this.renderDialog(context)}
                    </>
                )}
            </LocalizationContextConsumer>
        );
    }

    private readonly renderDialog = (context: LocalizationContextInterface) => {
        return (
            <Dialog
                hidden={!this.state.showPayloadDialog}
                onDismiss={this.hidePayloadDialog}
                modalProps={{
                    className: 'delete-dialog',
                    isBlocking: false,
                }}
            >
                <Editor
                    language="json"
                    ref={payloadRef}
                    options={{
                        automaticLayout: true,
                        readOnly: true,
                    }}
                    height="275px"
                    value={JSON.stringify(this.state.payloadPreviewData, null, '\t')}
                />
                <DialogFooter>
                    <PrimaryButton
                        onClick={this.copyPayload}
                        text={context.t(ResourceKeys.common.copyableMaskField.copy.label)}
                    />
                </DialogFooter>
            </Dialog>
        );
    }

    private readonly createForm = (context: LocalizationContextInterface) => {
        if (this.state.parseMapTypeError || !this.props.settingSchema) { // Not able to parse interface definition, render raw json in editor instead
            return this.createJsonEditor(context);
        }
        else {
            return (
                <ErrorBoundary error={context.t(ResourceKeys.errorBoundary.text)}>
                    <Form
                        className="value-section"
                        formData={this.state.formData}
                        liveValidate={true}
                        onChange={this.onChangeForm}
                        schema={this.props.settingSchema as any} // tslint:disable-line: no-any
                        showErrorList={false}
                        uiSchema={{'ui:description': this.props.settingSchema.description, 'ui:disabled': false}}
                        widgets={fabricWidgets}
                        {...fabricFields}
                        ObjectFieldTemplate={ObjectTemplate}
                    >
                        {this.createActionButtons(context)}
                    </Form>
                </ErrorBoundary>
            );
        }
    }

    private readonly createJsonEditor = (context: LocalizationContextInterface) => {
        return (
            <form className="json-editor">
                <LabelWithTooltip
                    tooltipText={context.t(ResourceKeys.notifications.interfaceSchemaNotSupported, {
                        schema: this.props.schema
                    })}
                >
                    {context.t(ResourceKeys.deviceContent.value)}
                </LabelWithTooltip>
                <Editor
                    language="json"
                    options={{
                        automaticLayout: true,
                        readOnly: false
                    }}
                    height="30vh"
                    value={this.state.stringifiedFormData}
                    onChange={this.onChangeEditor}
                />
                {this.createActionButtons(context)}
            </form>
        );
    }

    private readonly onChangeForm = (data: any) => { // tslint:disable-line: no-any
        this.setState({formData: data.formData});
    }

    private readonly onChangeEditor = (data: string) => {
        this.setState({stringifiedFormData: data});
    }

    private readonly createPayloadPreview = () => {
        const payload = (this.state.parseMapTypeError || !this.props.settingSchema) ?
        this.state.stringifiedFormData && JSON.parse(this.state.stringifiedFormData) :
        dataToTwinConverter(this.state.formData, this.props.settingSchema, this.state.originalFormData).twin;
        this.setState({
            payloadPreviewData: this.props.craftPayload(payload || null),
            showPayloadDialog: true
        });
    }

    private readonly hidePayloadDialog = () => {
        this.setState({showPayloadDialog: false});
    }

    private readonly copyPayload = () => {
        const editor = payloadRef.current.editor;
        editor.setSelection(editor.getVisibleRanges()[0]);
        editor.focus();
        document.execCommand('copy');
    }

    private readonly createActionButtons = (context: LocalizationContextInterface) => {
        const payload = (this.state.parseMapTypeError || !this.props.settingSchema) ?
            this.state.stringifiedFormData && JSON.parse(this.state.stringifiedFormData) :
            dataToTwinConverter(this.state.formData, this.props.settingSchema, this.state.originalFormData).twin;
        return (
            <>
                <PrimaryButton
                    className="submit-button"
                    onClick={this.props.handleSave(payload)}
                    text={context.t(this.props.buttonText)}
                    iconProps={{ iconName: SUBMIT }}
                />
                <DefaultButton
                    className="preview-payload-button"
                    onClick={this.createPayloadPreview}
                    text={context.t(ResourceKeys.deviceSettings.previewPayloadButtonText)}
                    iconProps={{ iconName: CODE }}
                />
            </>
        );
    }
}
