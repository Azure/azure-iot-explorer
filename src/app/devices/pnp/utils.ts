/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { getDeviceIdFromQueryString, getModuleIdentityIdFromQueryString } from '../../shared/utils/queryStringHelper';
import { ROUTE_PARAMS } from '../../constants/routes';
import { getDeviceTwinAction, getModuleTwinAction } from './actions';
import { ModelDefinition } from '../../api/models/modelDefinition';
import { ComponentToModelId, JsonSchemaAdaptor } from '../../shared/utils/jsonSchemaAdaptor';
import { DEFAULT_COMPONENT_FOR_DIGITAL_TWIN } from '../../constants/devices';

export const getBackUrl = (path: string, search: string) => {
    const deviceId = getDeviceIdFromQueryString(search);
    const moduleId = getModuleIdentityIdFromQueryString(search);
    let url = `${path}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`;
    if (moduleId) {
        url += `&${ROUTE_PARAMS.MODULE_ID}=${encodeURIComponent(moduleId)}`;
    }
    return url;
};

export const dispatchGetTwinAction = (search: string, dispatch: (action: any) => void) => {  // tslint:disable-line:no-any
    const deviceId = getDeviceIdFromQueryString(search);
    const moduleId = getModuleIdentityIdFromQueryString(search);
    if (moduleId) {
        dispatch(getModuleTwinAction.started({deviceId, moduleId}));
    }
    else {
        dispatch(getDeviceTwinAction.started(deviceId));
    }
};

export const getComponentNameAndInterfaceIdArray = (modelDefinition: ModelDefinition): ComponentToModelId[] => {
    if (!modelDefinition) {
        return [];
    }
    const jsonSchemaAdaptor = new JsonSchemaAdaptor(modelDefinition);
    const components = jsonSchemaAdaptor.getComponentNameToModelIdMapping();
    // check if model contains no-component items
    if (jsonSchemaAdaptor.getNonWritableProperties().length +
        jsonSchemaAdaptor.getWritableProperties().length +
        jsonSchemaAdaptor.getCommands().length +
        jsonSchemaAdaptor.getTelemetry().length > 0) {
        components.unshift({
            componentName: DEFAULT_COMPONENT_FOR_DIGITAL_TWIN,
            modelId: modelDefinition['@id']
        });
    }
    return components;
};
