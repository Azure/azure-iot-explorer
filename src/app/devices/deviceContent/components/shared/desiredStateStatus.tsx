/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Icon } from 'office-ui-fabric-react/lib/Icon';
import { DesiredState } from '../../../../api/models/digitalTwinModels';
import LabelWithTooltip from '../../../../shared/components/labelWithTooltip';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../..//shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { ACCEPT, WARNING, SYNCH } from '../../../../constants/iconNames';
import { DesiredStateStatus } from '../../../../constants/shared';

export const RenderDesiredState = (state: DesiredState) => {
    return (
        <LocalizationContextConsumer>
            {(context: LocalizationContextInterface) => (
                <LabelWithTooltip
                    tooltipText={state.description}
                >
                 {RenderStatusIcon(state.code, context)}
                </LabelWithTooltip>
            )}
        </LocalizationContextConsumer>
    );
};

export const RenderStatusIcon = (code: number, context: LocalizationContextInterface) => {
    switch (code) {
        case DesiredStateStatus.Success: return (
            <span className="status-success">
                <Icon className="status-icon" iconName={ACCEPT} />
                {context.t(ResourceKeys.deviceSettings.desiredState.success)}
            </span>
        );
        case DesiredStateStatus.Synching: return (
            <span className="status-synching">
                <Icon className="status-icon" iconName={SYNCH} />
                {context.t(ResourceKeys.deviceSettings.desiredState.synching)}
            </span>
        );
        case DesiredStateStatus.Error: return (
            <span className="status-error">
                <Icon className="status-icon" iconName={WARNING} />
                {context.t(ResourceKeys.deviceSettings.desiredState.error)}
            </span>
        );
        default: return (
            <span className="status-unknown">
                {context.t(ResourceKeys.deviceSettings.desiredState.unknown, {code})}
            </span>
        );
    }
};
