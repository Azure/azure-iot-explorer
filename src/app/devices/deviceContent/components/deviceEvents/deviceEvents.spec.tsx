/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { TextField } from 'office-ui-fabric-react/lib/TextField';
import { CommandBar } from 'office-ui-fabric-react/lib/CommandBar';
import DeviceEventsComponent, { DeviceEventsState } from './deviceEvents';
import { mountWithLocalization, testSnapshot } from '../../../../shared/utils/testHelpers';
import { AppEnvironment } from '../../../../constants/shared';

describe('components/devices/deviceEvents', () => {

    const pathname = `#/devices/detail/events/?id=device1`;

    const location: any = { // tslint:disable-line:no-any
        pathname
    };

    const routerProps: any = { // tslint:disable-line:no-any
        history: {
            location
        },
        location,
        match: {}
    };

    const getComponent = (overrides = {}) => {
        const props = {
            connectionString: '',
            ...routerProps,
            ...overrides,
        };
        return (<DeviceEventsComponent {...props} />);
    };

    it('matches snapshot in electron', () => {
        process.env = {...process.env, NODE_ENV: AppEnvironment.ProdElectron};
        const wrapper = getComponent();
        testSnapshot(wrapper);
    });

    it('matches snapshot in hosted environment', () => {
        process.env = {...process.env, NODE_ENV: AppEnvironment.ProdHosted};
        const wrapper = getComponent();
        testSnapshot(wrapper);
    });

    it('changes state accordingly when command bar buttons are clicked', () => {
        const wrapper = mountWithLocalization(getComponent());
        const commandBar = wrapper.find(CommandBar).first();
        // click the start button
        commandBar.props().items[0].onClick(null);
        wrapper.update();
        expect((wrapper.state() as DeviceEventsState).hasMore).toBeTruthy();

        // click the start button again which has been toggled to stop button
        commandBar.props().items[0].onClick(null);
        wrapper.update();
        expect((wrapper.state() as DeviceEventsState).hasMore).toBeFalsy();

        // clear events button should be disabled
        expect(commandBar.props().items[1].disabled).toBeTruthy();

        // click the show system property button
        commandBar.props().items[2].onClick(null); // tslint:disable-line:no-magic-numbers
        wrapper.update();
        expect((wrapper.state() as DeviceEventsState).showSystemProperties).toBeTruthy();
    });

    it('changes state accordingly when consumer group value is changed', () => {
        const wrapper = mountWithLocalization(getComponent());
        const textField = wrapper.find(TextField).first();
        textField.instance().props.onChange({ target: null}, 'testGroup');
        wrapper.update();
        expect((wrapper.state() as DeviceEventsState).consumerGroup).toEqual('testGroup');
    });

    it('renders events', () => {
        const wrapper = mountWithLocalization(getComponent());
        const events = [{
            body: {
                humid: 123
            },
            enqueuedTime: '2019-10-14T21:44:58.397Z',
            properties: {
              'iothub-message-schema': 'humid'
            }
        }];
        wrapper.setState({events});
        wrapper.update();
        const enqueueTime = wrapper.find('h5');
        // tslint:disable-next-line:no-any
        expect((enqueueTime.props().children as any).join('')).toBeDefined();

        // click the clear events button
        const commandBar = wrapper.find(CommandBar).first();
        commandBar.props().items[1].onClick(null);
        wrapper.update();
        expect((wrapper.state() as DeviceEventsState).events).toEqual([]);
    });
});
