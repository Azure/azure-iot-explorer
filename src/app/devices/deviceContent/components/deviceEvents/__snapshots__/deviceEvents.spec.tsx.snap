// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/devices/deviceEvents matches snapshot 1`] = `
<div
  className="device-events"
  key="device-events"
>
  <StyledCommandBarBase
    className="command"
    items={
      Array [
        Object {
          "ariaLabel": "deviceEvents.command.start",
          "disabled": false,
          "iconProps": Object {
            "iconName": "Play",
          },
          "key": "Play",
          "name": "deviceEvents.command.start",
          "onClick": [Function],
        },
        Object {
          "ariaLabel": "deviceEvents.command.clearEvents",
          "disabled": true,
          "iconProps": Object {
            "iconName": "Clear",
          },
          "key": "Clear",
          "name": "deviceEvents.command.clearEvents",
          "onClick": [Function],
        },
        Object {
          "ariaLabel": "deviceEvents.command.showSystemProperties",
          "disabled": false,
          "iconProps": Object {
            "iconName": "Checkbox",
          },
          "key": "CheckboxComposite",
          "name": "deviceEvents.command.showSystemProperties",
          "onClick": [Function],
        },
      ]
    }
  />
  <h3>
    deviceEvents.headerText
  </h3>
  <StyledTextFieldBase
    ariaLabel="deviceEvents.consumerGroups.label"
    className="consumer-group-text-field"
    disabled={false}
    label="deviceEvents.consumerGroups.label"
    onChange={[Function]}
    onRenderLabel={[Function]}
    underlined={true}
    value="$Default"
  />
  <InfiniteScroll
    className="device-events-container"
    element="div"
    getScrollParent={null}
    hasMore={false}
    initialLoad={true}
    isReverse={true}
    key="scroll"
    loadMore={[Function]}
    loader={
      <div
        className="events-loader"
      >
        <StyledSpinnerBase />
        <h4>
          deviceEvents.infiniteScroll.loading
        </h4>
      </div>
    }
    pageStart={0}
    ref={null}
    role="main"
    threshold={250}
    useCapture={false}
    useWindow={true}
  >
    <div
      className="scrollable-sm"
    />
  </InfiniteScroll>
</div>
`;
